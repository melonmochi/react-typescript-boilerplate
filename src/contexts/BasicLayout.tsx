import React, { FC, createContext, useReducer } from "react";
import { BasicLayout as BasicLayoutType } from "typings";
import { basicLayoutReducers as reducers } from "@/reducers";

const initialState: BasicLayoutType.State = {
  collapsed: false,
  settings: {
    colorWeak: false,
    contentWidth: "Fluid",
    fixSiderbar: false,
    fixedHeader: false,
    layout: "sidemenu",
    menu: { locale: true },
    navTheme: "dark",
    title: "React TypeScript"
  }
};

export const Context = createContext<BasicLayoutType.Interface>({
  state: initialState,
  dispatch: () => {}
});

export const ContextProvider: FC = props => {
  const [state, dispatch] = useReducer(reducers, initialState);
  return (
    <Context.Provider value={{ state, dispatch }}>
      {props.children}
    </Context.Provider>
  );
};
