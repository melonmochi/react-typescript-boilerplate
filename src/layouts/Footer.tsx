import React from "react";
import { Icon } from "antd";
import { DefaultFooter } from "@ant-design/pro-layout";

const Footer: React.FC = () => (
  <>
    <DefaultFooter
      copyright="2019 melonmochi"
      links={[
        {
          key: "React TypeScript Boilerplate",
          title: "React TypeScript Boilerplate",
          href: "https://github.com/melonmochi/react-typescript-boilerplate",
          blankTarget: true
        },
        {
          key: "github",
          title: <Icon type="github" />,
          href: "https://github.com/melonmochi/react-typescript-boilerplate",
          blankTarget: true
        },
        {
          key: "Paradigma Digital",
          title: "Paradigma Digital",
          href: "https://www.paradigmadigital.com/",
          blankTarget: true
        }
      ]}
    />
    <div
      style={{
        padding: "0px 24px 24px",
        textAlign: "center"
      }}
    >
      <a
        href="https://www.netlify.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <img
          src="https://www.netlify.com/img/global/badges/netlify-color-bg.svg"
          width="82px"
          alt="netlify logo"
        />
      </a>
    </div>
  </>
);

export default Footer;
