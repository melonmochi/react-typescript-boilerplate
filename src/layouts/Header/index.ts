export { default as Dropdown } from "./Dropdown";
export { default as LanguageDropdown } from "./LanguageDropdown";
export { default as RightContent } from "./RightContent";
