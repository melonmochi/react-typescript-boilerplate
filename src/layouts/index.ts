export { default as BasicLayout } from "./BasicLayout";
export { default as Footer } from "./Footer";
