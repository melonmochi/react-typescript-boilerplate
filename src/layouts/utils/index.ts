export { default as mapComponent } from "./mapComponent";
export { translateRoute } from "./translateRoute";
