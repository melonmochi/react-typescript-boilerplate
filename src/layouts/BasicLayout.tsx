import React, { FC, useContext } from "react";
import ProLayout, {
  SettingDrawerProps,
  BasicLayoutProps as ProLayoutProps
} from "@ant-design/pro-layout";
import { BasicLayoutContext, BasicLayoutContextProvider } from "@/contexts";
import logo from "../assets/logo.png";
import Footer from "./Footer";
import { Link } from "react-router-dom";
import { RightContent } from "./Header";

export interface BasicLayoutProps extends ProLayoutProps {
  settings?: SettingDrawerProps["settings"];
}

export const BasicLayout: FC<BasicLayoutProps> = props => {
  const { state, dispatch } = useContext(BasicLayoutContext);
  const { collapsed } = state;
  const { settings } = props;

  const footerRender: BasicLayoutProps["footerRender"] = () => <Footer />;

  const handleMenuCollapse = (payload: boolean): void => {
    dispatch({ type: "CHANGE_COLLAPSED", payload });
  };

  const menuHeaderRender: BasicLayoutProps["menuHeaderRender"] = (
    logoDom,
    titleDom
  ) => (
    <Link to="/">
      {logoDom}
      {titleDom}
    </Link>
  );

  const rightContentRender: BasicLayoutProps["rightContentRender"] = rightProps => (
    <RightContent {...rightProps} />
  );

  return (
    <>
      <ProLayout
        collapsed={collapsed}
        footerRender={footerRender}
        logo={logo}
        menuHeaderRender={menuHeaderRender}
        onCollapse={handleMenuCollapse}
        rightContentRender={rightContentRender}
        title="React TypeScript"
        {...settings}
      />
    </>
  );
};

export const ConnectedBasicLayout: FC = () => {
  return (
    <BasicLayoutContextProvider>
      <BasicLayout />
    </BasicLayoutContextProvider>
  );
};

export default ConnectedBasicLayout;
