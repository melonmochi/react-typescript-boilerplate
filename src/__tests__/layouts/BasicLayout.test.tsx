import * as React from "react";
import { shallow } from "enzyme";
import ProLayout from "@ant-design/pro-layout";
import { BasicLayout, ConnectedBasicLayout } from "@/layouts/BasicLayout";

describe("BasicLayout", () => {
  describe("Without context", () => {
    const container = shallow(<BasicLayout />);
    test("Snapshot", () => {
      expect(container).toMatchSnapshot();
    });

    describe("fires props actions", () => {
      describe("ProLayout", () => {
        const proLayout = container.find(ProLayout);
        test("onCollapse", () => {
          proLayout.props().onCollapse(true);
        });
        test("footerRender", () => {
          const footerRender = proLayout.props().footerRender;
          footerRender && footerRender({}, null);
        });
      });
    });
  });

  describe("With Context", () => {
    const basicLayoutWithContext = shallow(<ConnectedBasicLayout />);
    test("Snapshot", () => {
      expect(basicLayoutWithContext).toMatchSnapshot();
    });
  });
});
