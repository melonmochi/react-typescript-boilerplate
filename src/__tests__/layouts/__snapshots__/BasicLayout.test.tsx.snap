// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicLayout With Context Snapshot 1`] = `
<Component>
  <Component />
</Component>
`;

exports[`BasicLayout Without context Snapshot 1`] = `
<div>
  <BasicLayout
    breadcrumbRender={[Function]}
    collapsed={false}
    colorWeak={false}
    contentWidth="Fluid"
    fixSiderbar={false}
    fixedHeader={false}
    footerRender={[Function]}
    iconfontUrl=""
    itemRender={[Function]}
    layout="sidemenu"
    logo="test-file-stub"
    menu={
      Object {
        "locale": true,
      }
    }
    menuHeaderRender={[Function]}
    menuItemRender={[Function]}
    navTheme="dark"
    onCollapse={[Function]}
    primaryColor="daybreak"
    rightContentRender={[Function]}
    route={
      Object {
        "path": "/",
        "routes": Array [
          Object {
            "children": undefined,
            "component": <Welcome />,
            "icon": "smile",
            "name": "Welcome",
            "nameKey": "welcome.title",
            "path": "/welcome",
          },
          Object {
            "children": Array [
              Object {
                "children": undefined,
                "component": <NoFoundPage />,
                "icon": "warning",
                "name": "404",
                "nameKey": "global.404.title",
                "path": "/global/404",
              },
            ],
            "icon": "html5",
            "name": "Global",
            "nameKey": "global.title",
            "path": "/global",
          },
          Object {
            "children": Array [
              Object {
                "children": undefined,
                "component": <Unknown />,
                "icon": "cloud-server",
                "name": "List Table",
                "nameKey": "list.listTable",
                "path": "/list/list-table",
              },
              Object {
                "children": undefined,
                "component": <Unknown />,
                "icon": "solution",
                "name": "Custom List Table",
                "nameKey": "list.customListTable",
                "path": "/list/custom-list-table",
              },
            ],
            "icon": "unordered-list",
            "name": "List",
            "nameKey": "list.title",
            "path": "/list",
          },
        ],
      }
    }
    title="React TypeScript"
  />
  <SettingDrawer
    onSettingChange={[Function]}
    settings={
      Object {
        "colorWeak": false,
        "contentWidth": "Fluid",
        "fixSiderbar": false,
        "fixedHeader": false,
        "layout": "sidemenu",
        "menu": Object {
          "locale": true,
        },
        "navTheme": "dark",
        "title": "React TypeScript",
      }
    }
  />
</div>
`;
